### PR TITLE
Basic todo list CRUD app

### PR DESCRIPTION
**Pull request by vermakhushboo**
Added GET, POST, PUT and DELETE requests to a /todos route with task and is_complete properties.

### CODE REVIEW

File: app/server.php [hunk 1]
```diff
@@ -83,6 +84,77 @@ function($request, $response) {
...
+
+App::post('/todos')
+    ->inject('response')
+    ->param('task', "", new Wildcard(), 'Prefs key-value JSON object.')
+    ->param('is_complete', true, new Wildcard(), 'Tells whether task is complete or not')
+    ->action(
+        function($task, $is_complete, $response) {
+            $path = \realpath('/app/app/todos.json');
+            $data = json_decode(file_get_contents($path));
+            $task_entry = ['task' => $task, 'is_complete' => $is_complete];
```
Comment (christyjacob4) [inline app/server.php:106]: At this point, you can make use of the https://www.php.net/manual/en/function.uniqid.php function that PHP offers to generate unique IDs for each todo

File: app/server.php [hunk 2]
```diff
@@ -83,6 +84,77 @@ function($request, $response) {
...
+            array_push($data, $task_entry);
+            $jsonData = json_encode($data, JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES);
+            file_put_contents($path, $jsonData);
+            $response->json($jsonData);
```
Comment (christyjacob4) [inline app/server.php:110]:
```suggestion
            $response->json($data);
```

$jsonData is a json string and will cause an error. `$response->data` method expects an array as input.

File: app/server.php [hunk 3]
```diff
@@ -83,6 +84,77 @@ function($request, $response) {
...
+        }
+    );
+
+App::put('/todos/:todo_name')
```
Comment (christyjacob4) [inline app/server.php:114]:
```suggestion
App::put('/todos/:id')
```

Let's not use the todo_name to identify a todo.

File: app/server.php [hunk 4]
```diff
@@ -83,6 +84,77 @@ function($request, $response) {
...
+    ->inject('response')
+    ->param('todo_name', "", new Wildcard(), 'name of the todo')
```
Comment (christyjacob4) [inline app/server.php:116]:
```suggestion
    ->param('id', "", new Wildcard(), 'id of the todo')
    ->param('todo_name', "", new Wildcard(), 'name of the todo')
```

Let's also include the `id` as a param here

File: app/server.php [hunk 5]
```diff
@@ -83,6 +84,77 @@ function($request, $response) {
...
+    ->param('is_complete', true, new Wildcard(), 'Tells whether task is complete or not')
+    ->action(
+        function($todo_name, $is_complete, $response) {
```
Comment (christyjacob4) [inline app/server.php:119]:
```suggestion
        function($id, $todo_name, $is_complete, $response) {
```

File: app/server.php [hunk 6]
```diff
@@ -83,6 +84,77 @@ function($request, $response) {
...
+            $path = \realpath('/app/app/todos.json');
+            $data = json_decode(file_get_contents($path));
+            // var_dump($data);
+            foreach($data as $object){
+                if($object->task == $todo_name){
```
Comment (christyjacob4) [inline app/server.php:124]:
```suggestion
                if($object->id == $id){
```

File: app/server.php [hunk 7]
```diff
@@ -83,6 +84,77 @@ function($request, $response) {
...
+                    $object->is_complete = $is_complete;
```
Comment (christyjacob4) [inline app/server.php:125]:
```suggestion
                    $object->is_complete = $is_complete;
                    $object->task = $task;
```

File: app/server.php [hunk 8]
```diff
@@ -83,6 +84,77 @@ function($request, $response) {
...
+                }
+            }
+            // var_dump($data);
+            $jsonData = json_encode($data, JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES);
+            file_put_contents($path, $jsonData);
+            $response->json($data);
+        }
+    );
+
+App::delete('/todos/:todo_name')
```
Comment (christyjacob4) [inline app/server.php:136]:
```suggestion
App::delete('/todos/:id')
```

File: app/server.php [hunk 9]
```diff
@@ -83,6 +84,77 @@ function($request, $response) {
...
+    ->inject('response')
+    ->param('todo_name', "", new Wildcard(), 'name of the todo')
```
Comment (christyjacob4) [inline app/server.php:138]:
```suggestion
    ->param('id', "", new Wildcard(), 'id of the todo')
```

File: app/server.php [hunk 10]
```diff
@@ -83,6 +84,77 @@ function($request, $response) {
...
+    ->action(
+        function($todo_name, $response) {
```
Comment (christyjacob4) [inline app/server.php:140]:
```suggestion
        function($id, $response) {
```

File: app/server.php [hunk 11]
```diff
@@ -83,6 +84,77 @@ function($request, $response) {
...
+            $path = \realpath('/app/app/todos.json');
+            $data = json_decode(file_get_contents($path));
+            var_dump($todo_name);
+            foreach($data as $object => $item){
+                if($item->task == $todo_name){
```
Comment (christyjacob4) [inline app/server.php:145]:
```suggestion
                if($item->id == $id){
```

File: app/server.php [hunk 12]
```diff
@@ -83,6 +84,77 @@ function($request, $response) {
...
+                    var_dump($object);
+                    unset($data[$object]);
+                }
+            }
+            var_dump($data);
+            $jsonData = json_encode($data, JSON_PRETTY_PRINT | JSON_UNESCAPED_SLASHES);
+            file_put_contents($path, $jsonData);
+            $response->json($data);
```
Comment (christyjacob4) [inline app/server.php:153]: I think it's a convention with DELETE endpoints , to return the item that was deleted. Can we do that ?

File: app/todos.json [hunk 13]
```diff
@@ -0,0 +1,14 @@
+[
+    {
+        "task": "Create design document",
```
Comment (christyjacob4) [inline app/todos.json:3]: Let's add some `id` attributes to these todos 💪

**Review by christyjacob4 — CHANGES_REQUESTED** [review]

File: app/server.php (hunk 12 above)
Comment (kodumbeats) [reply app/server.php:157]: There's no hard rule here - I believe the following are appropriate:
- Respond with 204 and an empty response body (what we do in Appwrite)
- Respond with a 200 and the item that was deleted

[ref](https://stackoverflow.com/questions/25970523/restful-what-should-a-delete-response-body-contain)

File: app/server.php [hunk 14]
```diff
@@ -83,6 +84,80 @@ function($request, $response) {
...
+        }
+    );
+
+    App::delete('/todos/:id')
+    ->inject('response')
+    ->param('id', "", new Wildcard(), 'id of the todo')
+    ->action(
+        function($id, $response) {
+            $path = \realpath('/app/app/todos.json');
+            $data = json_decode(file_get_contents($path));
```
Comment (christyjacob4) [inline app/server.php:146]:
```suggestion
            $data = json_decode(file_get_contents($path), true);
```

File: app/server.php [hunk 15]
```diff
@@ -83,6 +84,80 @@ function($request, $response) {
...
+            var_dump($id);
+            foreach($data as $object => $item){
+                if($item->id == $id){
+                    var_dump($object);
+                    unset($data[$object]);
+                }
```
Comment (christyjacob4) [inline app/server.php:152]:
```suggestion
            foreach($data as $object => $item){
                if($item['id'] == $id){
                    unset($data[$object]);
                }
            }
```

File: app/server.php (hunk 14 above)
Comment (christyjacob4) [reply app/server.php:146]: We need to read the file contents as an associative array by passing the `true` flag to `json_decode`

File: app/server.php (hunk 12 above)
Comment (christyjacob4) [reply app/server.php:157]: @vermakhushboo  let's follow the appwrite convention in that case 🙂

**Comment by christyjacob4** [comment]
The PR was meant to make the review process easier. The review is complete and the PR can be closed 👍

**Pull request closed without merging**